plugins/pulp_rpm/plugins/importers/yum/upload.py

```diff
@@ -518,6 +531,54 @@ def _extract_rpm_data(type_id, rpm_filename):
     return rpm_data
 
 
+def _extract_drpm_data(drpm_filename):
+    """
+    Extract a dict of information for a given DRPM.
+
+    :param drpm_filename: full path to the package to analyze
+    :type  drpm_filename: str
+
+    :return: dict of data about the package
+    :rtype:  dict
+    """
+    drpm_data = dict()
+
+    headers = rpm_parse.package_headers_drpm(drpm_filename)
+
+    try:  # "handle" rpm-only drpms (without rpm header)
+        rpm_headers = rpm_parse.package_headers(drpm_filename)
+    except rpm.error:
+        raise RPMOnlyDRPMsAreNotSupported(drpm_filename)
+
+    drpm_data['signature'] = rpm_parse.package_signature(rpm_headers)
+    drpm_data['arch'] = rpm_headers['arch']
+
+    old_nevr = old_name, old_epoch, old_version, old_release = rpm_parse.nevr(headers["old_nevr"])
+    new_nevr = new_name, new_epoch, new_version, new_release = rpm_parse.nevr(headers["nevr"])
+
+    drpm_data['sequence'] = headers["old_nevr"] + "-" + headers["seq"]
+
+    drpm_data['epoch'] = str(new_epoch)
+    drpm_data['oldepoch'] = str(old_epoch)
+
+    drpm_data['version'] = str(new_version)
+    drpm_data['oldversion'] = str(old_version)
+
+    drpm_data['release'] = new_release
+    drpm_data['oldrelease'] = old_release
+
+    drpm_data['new_package'] = new_name
+    drpm_data['size'] = os.stat(drpm_filename)[stat.ST_SIZE]
+
+    old_evr = rpm_parse.nevr_to_evr(*old_nevr)
+    new_evr = rpm_parse.nevr_to_evr(*new_nevr)
+    drpm_data['filename'] = "drpms/%s-%s_%s.%s.drpm" % (new_name, rpm_parse.evr_to_str(*old_evr),
```

Comment:
If I recall correctly, I think the filename is in the unit key as an easy way to capture both the "before" and "after" versions of the package. I'm not sure why those fields weren't just explicitly added to the model, but who knows. Note that Fedora has DRPM files with names like ``python-2.7.11-4.fc24_2.7.12-2.fc24.x86_64.drpm``.

So I am inclined to enforce that at upload time to make sure we get uniqueness, and invite publishers of DRPMs to join that convention. In the rare case where someone uploads the same DRPM that they also got from a sync, they can live with dual storage.

Probably we should do the same enforcement during sync, but that whole workflow is due for a re-think. The way we model DRPMs is not quite right, and there's another redmine issue about that somewhere in our tracker.